core/primitives/src/shard_layout.rs

```diff
@@ -235,15 +225,210 @@ impl ShardLayoutV2 {
     }
 }
 
+// Custom serialization required for handling maps with non-string keys
+impl serde::Serialize for ShardLayoutV2 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
```

Comment:
I would be tempted to introduce a helper structure e.g. SerdeShardLayoutV2 that has the maps using strings as keys. Then the custom serialization could be converting to/from the `SerdeShardLayoutV2` and using it's serialization. WDYT?

Reply:
Sounds like a great idea if `SerdeShardLayoutV2` <-> `ShardLayoutV2 ` could make de/ser' simpler. Or is there any other reasons you want to do it this way?

Reply:
No other reasons, just to make de/ser simpler.
It's totally not because I don't want to review custom serialization ;)

Reply:
Have tried it out, it's indeed much simpler :)

Reply:
yeah, the new version is lovely, thanks!